src/services/SendEmailProcessor.ts

```diff
@@ -44,15 +49,62 @@ export class SendEmailProcessor {
 
 	async processRequest(eventBody: any): Promise<EmailResponse> {
 		const email = Email.parseRequest(JSON.stringify(eventBody.Message));
+		// Validate Email model
+		try {
+			await this.validationHelper.validateModel(email, this.logger);
+		} catch (error) {
+			this.logger.error("Failed to Validate Email model data", { messageCode: MessageCodes.MISSING_MANDATORY_FIELDS });
+			throw new AppError(HttpCodesEnum.SERVER_ERROR, "Failed to Validate Email model data.");
+		}
 
-		await this.validationHelper.validateModel(email, this.logger);
+		//Retrieve session event record for the userId
+		let session;
+		try {
+			session = await this.iprService.getSessionBySub(email.userId);
+			this.logger.debug("Session retrieved from session store");
+			if (!session) {
+				this.logger.error("No session event found for this userId", { messageCode: MessageCodes.SESSION_NOT_FOUND });
+				throw new AppError(HttpCodesEnum.SERVER_ERROR, "No session event found for this userId");
+			}
+
+		} catch (error) {
+			this.logger.error({ message: "getSessionByUserId - failed executing get from dynamodb:", error });
+			throw new AppError(HttpCodesEnum.SERVER_ERROR, "Error retrieving Session");
+		}
 
-		const emailResponse: EmailResponse = await this.govNotifyService.sendEmail(email);
+		// Validate the notified field is set to true
+		if (!session.notified) {
+			this.logger.error("Notified flag is not set to true for this user session event", { messageCode: MessageCodes.NOTIFIED_FLAG_NOT_SET_TO_TRUE });
+			throw new AppError(HttpCodesEnum.SERVER_ERROR, "Notified flag is not set to true for this user session event");
+		}
 
-		await this.iprService.sendToTXMA({
-			event_name: "IPR_RESULT_NOTIFICATION_EMAILED",
-			...buildCoreEventFields({ email: email.emailAddress }),
-		});
+		// Validate if the record is missing some fields related to the Events and log the details and do not notify the User.
+		try {
+			this.validationHelper.validateSessionEventFields(session);
```

Comment:
Just for my own understanding - why do we need to do these checks? If I understand correctly, it [looks like](https://github.com/alphagov/di-ipvreturn-api/blob/F2F-666/src/services/SessionEventProcessor.ts#L66-L67) the messages only get put on the queue to be picked up by this service if we're sure that all the checks have passed

Reply:
Its more to make the lambda more robust enough to handle any false/invalid messages that get sent over (via hacking or testing etc) which may end up triggering an email to the given emailId without any validations in place. Ex: By some means if the session event for a userId was updated (removal of some necessary field) and if a random message is placed in the queue for this userId, then without such validations in place, we may trigger the email out. Ideally the streamProcessor would not place this notify message for such scenario but its more to handle any random messages being pushed to our govNotify queue which we are trying to make it more defensive.

Reply:
I see, thanks for explaining. 

If we are reusing the logic do we want to extract it into a function? Just thinking that if we need to add another check we're going to have to add it in two places, and it could easily be missed. 

Also we have like 20 lines of code that are repeated!